src/channel/NodeChannel.js

```diff
@@ -18,73 +18,151 @@
  * ‍
  */
 
+import tls from "tls";
 import { Client, credentials } from "@grpc/grpc-js";
 import Channel from "./Channel.js";
 import GrpcServicesError from "../grpc/GrpcServiceError.js";
 import GrpcStatus from "../grpc/GrpcStatus.js";
 import { ALL_NETWORK_IPS } from "../constants/ClientConstants.js";
 
-/**
- * @property {?HashgraphProto.proto.CryptoService} _crypto
- * @property {?HashgraphProto.proto.SmartContractService} _smartContract
- * @property {?HashgraphProto.proto.FileService} _file
- * @property {?HashgraphProto.proto.FreezeService} _freeze
- * @property {?HashgraphProto.proto.ConsensusService} _consensus
- * @property {?HashgraphProto.proto.NetworkService} _network
- */
+/** @type {{ [key: string]: string }} */
+const certificateCache = {};
+/** @type {{ [key: string]: Client }} */
+const clientCache = {};
+
 export default class NodeChannel extends Channel {
     /**
      * @internal
      * @param {string} address
-     * @param {string=} cert
      * @param {number=} maxExecutionTime
      */
-    constructor(address, cert, maxExecutionTime) {
+    constructor(address, maxExecutionTime) {
         super();
 
-        this.cert = cert;
+        /** @type {Client | null} */
+        this._client = null;
+
+        this.address = address;
         this.maxExecutionTime = maxExecutionTime;
 
+        const { ip, port } = this.parseAddress(address);
+        this.nodeIp = ip;
+        this.nodePort = port;
+    }
+
+    /**
+     * Convert certificate bytes to PEM format
+     * @param {Buffer} certBytes
+     * @returns {string}
+     */
+    bytesToPem(certBytes) {
+        const base64Cert = certBytes.toString("base64");
+        const lines = base64Cert.match(/.{1,64}/g)?.join("\n") || "";
+        return `-----BEGIN CERTIFICATE-----\n${lines}\n-----END CERTIFICATE-----`;
+    }
+
+    /**
+     * Validates and parses an address in the "IP:Port" format.
+     * @param {string} address
+     * @returns {{ ip: string, port: string }}
+     */
+    parseAddress(address) {
+        const [ip, port] = address.split(":");
+        if (!ip || !port) {
+            throw new Error(
+                "Invalid address format. Expected format: 'IP:Port'",
+            );
+        }
+        return { ip, port };
+    }
+
+    /**
+     * Retrieve the server's certificate dynamically.
+     * @param {string} address
+     * @returns {Promise<string>}
+     */
+    async _retrieveCertificate(address) {
+        if (certificateCache[address]) {
+            return certificateCache[address];
```

Comment:
if we have the cert in the cache, it will not return a Promise, but the String?

Reply:
Yep, that's correct.